h2o-core/src/main/java/water/parser/CsvParser.java

```diff
@@ -459,17 +460,20 @@
   }
 
   @Override protected int fileHasHeader(byte[] bits, ParseSetup ps) {
-    boolean hasHdr = true;
+    boolean hasHdr = false;
     String[] lines = getFirstLines(bits);
     if (lines != null && lines.length > 0) {
       String[] firstLine = determineTokens(lines[0], _setup._separator, _setup._single_quotes);
-      if (_setup._column_names != null) {
-        for (int i = 0; hasHdr && i < firstLine.length; ++i)
+      if (_setup._column_names != null && _setup._column_names.length == firstLine.length) {
+        for (int i = 0; !hasHdr && i < firstLine.length; ++i)
           hasHdr = (_setup._column_names[i] == firstLine[i]) || (_setup._column_names[i] != null && _setup._column_names[i].equalsIgnoreCase(firstLine[i]));
       } else { // declared to have header, but no column names provided, assume header exist in all files
+        hasHdr = true;
         _setup._column_names = firstLine;
       }
-    } // else FIXME Throw exception
```

Comment:
Why do you need to throw the exception?

Reply:
Because the given file(s) is/are empty and there is no way to decide if there are headers or not ? At least that is what this method's contract promises.

Reply:
If it is empty, it doesn't have headers in my opinion. That was also the previous behavior.

Reply:
Fixed.